### PR TITLE
New Access workflow for Mattermost plugin

### LINKED ISSUE #187: Split out Enterprise/OSS Mattermost integration
Similar to #183, implement Mattermost enterprise (readonly, DMs) and OSS (shared channel notification + notification-only mode) integrations. 

### Mattermost Enterprise

- [ ] Implement Mattermost users mapping to Teleport users mapping
- [ ] Send Mattermost DMs to corresponding users for new access requests, those DMs should contain links to Teleport Web UI
- [ ] Send Access Request notification when a request is approved or denied to a shared channel on Mattermost

### Mattermost OSS
- [ ] Support notification-only (readonly) mode for Mattermost integration
- [ ] Support access request notes in the notification body (like in #184)

### Documentation

- [ ] Review and update the documentation on the website
- [ ] Review and update docs in install.md

### PR DESCRIPTION
**Pull request by marshall-lee**
Closes #187
Related to #196

### CODE REVIEW

**Comment by klizhentas** [comment]
@fspmarshall @r0mant I know you folks are very busy, but need your review on mattermost plugin. We have to deliver it by April 7th release.

**Review by fspmarshall — APPROVED** [review]
I've never used mattermost so I can't really comment on the workflow itself, but code looks good to me :)

File: access/mattermost/bot.go [hunk 1]
```diff
@@ -2,204 +2,302 @@ package main
...
+			Timeout: mmHTTPTimeout,
+			Transport: &http.Transport{
+				MaxConnsPerHost:     mmMaxConns,
+				MaxIdleConnsPerHost: mmMaxConns,
+			},
+		}).
+		SetHostURL(conf.URL).
+		SetHeader("Content-Type", "application/json").
+		SetHeader("Accept", "application/json").
+		SetHeader("Authorization", "BEARER "+conf.Token)
```
Comment (r0mant) [inline access/mattermost/bot.go:89]: Nit: Don't think the case matters but I've never seen it capitalized, `Bearer`.

File: access/mattermost/bot.go [hunk 2]
```diff
@@ -2,204 +2,302 @@ package main
...
+}
+
+type etagCacheCtxKey struct{}
+
+type etagCacheEntry struct {
+	etag  string
+	value interface{}
+}
+
+func NewBot(conf MattermostConfig, clusterName, webProxyAddr string) (Bot, error) {
```
Comment (r0mant) [inline access/mattermost/bot.go:67]: Nit: Add godoc.

File: access/mattermost/bot.go [hunk 3]
```diff
@@ -2,204 +2,302 @@ package main
...
+			return nil
+		}
+
+		etag := resp.Header().Get("ETag")
+		if etag == "" {
+			return nil
+		}
+
+		if resp.IsSuccess() || resp.StatusCode() == http.StatusNotModified {
+			cache.Add(cacheKey, etagCacheEntry{etag: etag, value: resp.Result()})
```
Comment (r0mant) [inline access/mattermost/bot.go:155]: Is `cache` goroutine-safe?

File: access/mattermost/bot.go [hunk 4]
```diff
@@ -2,204 +2,302 @@ package main
...
+	})
+
+	return Bot{
+		client:      client,
+		clusterName: clusterName,
+		webProxyURL: webProxyURL,
+	}, nil
+}
+
+func (b Bot) HealthCheck(ctx context.Context) error {
```
Comment (r0mant) [inline access/mattermost/bot.go:168]: Godoc.

File: access/mattermost/bot.go [hunk 5]
```diff
@@ -2,204 +2,302 @@ package main
...
+	_, err := b.GetMe(ctx)
+	return err
 }
 
-func (b *Bot) HealthCheck() error {
-	_, resp := b.client.GetTeamByName(b.team, "")
-	if resp.Error != nil {
-		return trace.Wrap(resp.Error)
+func (b Bot) GetMe(ctx context.Context) (User, error) {
```
Comment (r0mant) [inline access/mattermost/bot.go:173]: Godoc.

File: access/mattermost/bot.go [hunk 6]
```diff
@@ -2,204 +2,302 @@ package main
...
+			errors = append(errors, trace.Wrap(err))
 		}
-		actions = append(actions, approveAction, denyAction)
 	}
 
-	text, err := b.buildPostText(reqID, reqData, status)
-	if err != nil {
-		return nil, err
+	if len(errors) > 0 {
+		return errors[0]
```
Comment (r0mant) [inline access/mattermost/bot.go:294]: I'd `return trace.NewAggregate(errors...)` instead.

**Review by r0mant — APPROVED** [review]

File: access/mattermost/bot.go (hunk 3 above)
Comment (marshall-lee) [reply access/mattermost/bot.go:155]: Yes it uses a mutex internally

**Review by klizhentas — CHANGES_REQUESTED** [review]
I've tested the plugin and it works! Can you please fix a couple of issues before we move forward:

* [ ] Formatting error in error message:

```
WARN   Failed to notify a suggested reviewer: %q does not look like a valid emailalice request_id:a9770b7a-d6ab-4c43-9e58-c17a3a0f017c request_op:put mattermost/app.go:269
```

* [x] The message from bot should contain the URL of the request and link should be https://teleport:port/web/requests/request-id

**Pull request merged**